include/boost/gil/image_processing/morphology.hpp

```diff
@@ -0,0 +1,282 @@
+#define BOOST_GIL_IMAGE_PROCESSING_MORPHOLOGY_HPP
+#include <boost/gil/extension/numeric/kernel.hpp>
+#include <boost/gil/gray.hpp>
+#include <boost/gil/image_processing/threshold.hpp>
+
+namespace boost {
+namespace gil {
+namespace detail {
+  enum class morphological_operation {
+  dilation,
+  erosion,
+};
+/// \addtogroup ImageProcessing
+/// @{
+
+/// \brief Implements morphological operations at pixel level.This function
+/// compares neighbouring pixel values according to the kernel and choose
+/// minimum/mamximum neighbouring pixel value and assigns it to the pixel under
+/// consideration. \param src_view - Source/Input image view. \param dst_view -
+/// View which stores the final result of operations performed by this function.
+/// \param kernel - Kernel matrix/structuring element containing 0's and 1's
+/// which will be used for applying the required morphological operation. \param
+/// identifier - Indicates the type of morphological operation to be applied.
+/// \tparam SrcView type of source image.
+/// \tparam DstView type of output image.
+/// \tparam Kernel type of structuring element.
+template <typename SrcView, typename DstView, typename Kernel>
+void morph_impl(SrcView const &src_view, DstView const &dst_view,
+                Kernel const &kernel, morphological_operation identifier) {
+  std::ptrdiff_t flip_ker_row, flip_ker_col, row_boundary, col_boundary;
+  typename channel_type<typename SrcView::value_type>::type target_element;
+  for (std::ptrdiff_t view_row = 0; view_row < src_view.height(); ++view_row) {
+    for (std::ptrdiff_t view_col = 0; view_col < src_view.width(); ++view_col) {
+      target_element = src_view(view_col, view_row);
+      for (std::size_t kernel_row = 0; kernel_row < kernel.size();
+           ++kernel_row) {
+        flip_ker_row =
+            kernel.size() - 1 - kernel_row; // row index of flipped kernel
+
+        for (std::size_t kernel_col = 0; kernel_col < kernel.size();
+             ++kernel_col) {
+          flip_ker_col =
+              kernel.size() - 1 - kernel_col; // column index of flipped kernel
+              
+            // We ensure that we consider only those pixels which are overlapped
+            // on a non-zero kernel_element as
+            if(kernel.at(flip_ker_row,flip_ker_col) == 0){
+              continue;
+            }
+          // index of input signal, used for checking boundary
+          row_boundary = view_row + (kernel.center_y() - flip_ker_row);
+          col_boundary = view_col + (kernel.center_x() - flip_ker_col);
+
+          // ignore input samples which are out of bound
+          if (row_boundary >= 0 && row_boundary < src_view.height() &&
+              col_boundary >= 0 && col_boundary < src_view.width()) {
+            
+            if (identifier == morphological_operation::dilation) {
+              if (src_view(col_boundary, row_boundary) > target_element)
+                target_element = src_view(col_boundary, row_boundary);
```

Comment:
Can we use std::max here? Lets try this:

```
using std::max;
target_element[0] = max(target_element[0], src_view(col_boundary, row_boundary)[0]);
```

Reply:
I can use std::max, but then I will have to typecast first and then compare like this
```
target_element = std::max((int)src_view(col_boundary, row_boundary), (int)target_element);
```
I am not sure what you meant by using an indexing operator in 
```
target_element[0] = max(target_element[0], src_view(col_boundary, row_boundary)[0]);
```
As per my understanding, neither target_element nor src_view(col_boundary, row_boundary) is an array, please correct me if I am wrong.

Reply:
@simmplecoder ,I have formatted my files with your version of ```.clang-format``` . Please review those changes and tell me which version would you like me to use for the above point. I have simply commented my earlier approach, I will replace that with the final version in my next commit .

Reply:
Also, I will handle the merge conflict in my next commit .

Reply:
It doesn't have to be array. Homogenous pixels (where all channels are of the same type) have overloaded `operator[]` so you can access channels by index. Casting everything to `int` is wrong (examples: `gil::gray32f_image_t`, `gil::gray32_image_t`).

Reply:
Also, I'd suggest to use `(std::max)(a, b)` in case the `std::max(a, b)` fails with the annoying compilation errors from MSVC

Reply:
In this case, std::max(a, b) seems to work, your comment made me look deep into concepts like ADL(Argument dependent lookup) . Thanks for this great find :)

Reply:
Okay I will do that. I found a good reference on this [here](https://stackoverflow.com/questions/1632145/use-of-min-and-max-functions-in-c/2125302#2125302) . Thanks @mloskot for this great find.

Reply:
@mloskot , I found [this](https://dustri.org/b/min-and-max-macro-considered-harmful.html) blog and hence I would like to confirm this, which version would be more safe/suitable for our cause 
```
template<typename T>
inline T MAX(T a, T b) { return((a) > (b) ? a : b); }
```
or 
```
#define MAX(a,b) ((a) > (b) ? a : b)
```

Reply:
@meshtag  TBH, none of the two `MAX`-es above. Any reason why it can not be just the *existing* C++ *function* `std::max`?

Reply:
@mloskot , no specific reason in particular. Perhaps I misinterpreted your earlier comment. I thought you wanted me to disable ADL(Argument dependent lookup) here by using ```(std::max)(a, b)```. Since that doesn't seems to be the case now, I will keep it as it is.

Reply:
@meshtag Yes, I suggested that in order to avoid the potential of clashing with the `max` *macro* when compiling using the MSVC, but I did not suggest introducing new `MAX` macro or function, just to stick `(std::max)(...)` instead of `std::max(...)`.

Reply:
okay, pushed the code with changes.